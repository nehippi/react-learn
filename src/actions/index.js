export const addImages = (data) => ({type: 'ADD', data: data})
export const deleteImage = () => ({type: 'DELETE'})
export const updateCountToDelete = (countToDelete) => ({type: 'UPDATE_COUNT', countToDelete: countToDelete})
export const startDowloadingImages = () => ({type: 'START'})
export const endDowloadingImages = () => ({type: 'END'})
