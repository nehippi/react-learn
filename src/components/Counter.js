import React from "react";
import {connect} from "react-redux";
import PropTypes from 'prop-types';

export function Counter(props) {
    return (
        <p>Count of cats: {props.count}</p>
    );
}

const mapStateToProps = state => ({
    count: state.images.length
})

Counter.propTypes = {
    count: PropTypes.number
}

Counter.defaultProps={
    count: 0
}

export default connect(mapStateToProps)(Counter);
