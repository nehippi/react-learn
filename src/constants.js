export const IMAGE_HEIGHT =200
export const ADD ="ADD"
export const DELETE ="DELETE"
export const UPDATE_COUNT ="UPDATE_COUNT"
export const START ="START"
export const END ="END"
