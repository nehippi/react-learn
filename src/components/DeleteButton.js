import React from "react";
import {deleteImage} from "../actions";
import {connect} from "react-redux";
import PropTypes from 'prop-types';

export function DeleteButton(props) {
    const deleteCat = () => {
        const countToDelete = props.countToDelete;
        if (countToDelete !== 0) {
            props.deleteImages();
        }
    }
    return (<button onClick={deleteCat}>delete {props.countToDelete} cats</button>)
}

const mapStateToProps = state => ({
    countToDelete: state.countToDelete,
})

const mapDispatchToProps = dispatch => ({
    deleteImages: () => dispatch(deleteImage())
})

DeleteButton.defaultProps = {
    countToDelete: 0,
}

DeleteButton.propTypes = {
    countToDelete: PropTypes.number,
    deleteImages: PropTypes.func.isRequired,
}

export default connect(mapStateToProps, mapDispatchToProps)(DeleteButton)
