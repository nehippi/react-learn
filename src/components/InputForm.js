import React from "react";
import {updateCountToDelete} from "../actions";
import {connect} from "react-redux";
import PropTypes from "prop-types";

export function InputForm(props) {
    const handleChange = (e) => {
        props.updateCountToDelete(e.target.value);
    }

    return (
        <fieldset>
            <label>Enter count</label>
            <input type="number" value={props.countToDelete}
                   onChange={handleChange} min={0}/>
        </fieldset>
    );
}

const mapStateToProps = state => ({
    countToDelete: state.countToDelete
})

const mapDispatchToProps = dispatch => ({
    updateCountToDelete: countToDelete => dispatch(updateCountToDelete(countToDelete))
})

InputForm.defaultProps = {
    countToDelete: 0,
}

InputForm.propTypes = {
    countToDelete: PropTypes.number,
    updateCountToDelete: PropTypes.func.isRequired,
}

export default connect(mapStateToProps, mapDispatchToProps)(InputForm);
