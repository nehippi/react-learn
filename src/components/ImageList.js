import React from "react";
import PropTypes from "prop-types";
import {IMAGE_HEIGHT} from "../constants";


class ImageList extends React.Component {
    render() {
        const cats = this.props.data.map((data) => {
            return (
                <div key={data.id}>
                    <img src={data.url} height={IMAGE_HEIGHT} alt="could not download"/>
                </div>)
        });
        return (
            <div>
                {cats}
            </div>
        );
    }
}

ImageList.defaultProps = {
    data: [],
}

ImageList.propTypes = {
    data: PropTypes.array,
}

export default ImageList

