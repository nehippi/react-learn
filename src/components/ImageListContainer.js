import React from "react";
import {addImages} from "../actions";
import {connect} from "react-redux";
import ImageList from "./ImageList";
import PropTypes from "prop-types";

const IMAGE_HEIGHT = 200;

class ImageListContainer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isFetching: false,
        };
    }

    handleScroll = (e) => {
        const bottom = e.target.scrollingElement.scrollHeight - IMAGE_HEIGHT - e.target.scrollingElement.scrollTop <= e.target.scrollingElement.clientHeight + 1;
        if (bottom) {
            this.addCat();
        }
    }

    componentDidMount() {
        this.addCat();
        window.addEventListener("scroll", this.handleScroll);
    }

    componentWillUnmount() {
        window.removeEventListener("scroll", this.handleScroll);
    }

    addCat = () => {
        if (!this.state.isFetching) {
            this.setState({
                isFetching: true
            });
            fetch("https://api.thecatapi.com/v1/images/search?limit=10")
                .then((response) => {
                    return response.json();
                })
                .then((arrayOfData) => {
                    this.props.addImagesToStore(arrayOfData)
                    this.setState({
                        isFetching: false
                    });
                });

        }
    }

    render() {
        return (
            <ImageList data={this.props.images}/>
        );
    }

}

function mapStateToProps(state) {
    return {
        images: state.images,
    }
}

function mapDispatchToProps(dispatch) {
    return {
        addImagesToStore: (data) => dispatch(addImages(data)),
    }
}

ImageListContainer.defaultProps = {
    images: [],
}

ImageListContainer.propTypes = {
    images: PropTypes.array,
    addImagesToStore: PropTypes.func.isRequired,
}

export default connect(mapStateToProps, mapDispatchToProps)(ImageListContainer)