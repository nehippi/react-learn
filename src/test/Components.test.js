import React from 'react';
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Counter} from "../components/Counter";
import {DeleteButton} from "../components/DeleteButton";
import ImageList from "../components/ImageList";
import {ImageListContainer} from "../components/ImageListContainer";
import {InputForm} from "../components/InputForm";
import {ADD, END, START} from "../constants";
import {addCat} from "../actions";

configure({adapter: new Adapter()});


describe("Counter", () => {
    it("should render counter", () => {
        const wrapper = shallow(<Counter count={1}/>);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Delete button render", () => {
    it("should render Delete button", () => {
        const deleteImages = jest.fn();
        const wrapper = shallow(<DeleteButton countToDelete={1} deleteImages={deleteImages}/>);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Delete button onClick", () => {
    it("should handle onClick in Delete button", () => {
        const deleteImages = jest.fn();
        const wrapper = shallow(<DeleteButton countToDelete={1} deleteImages={deleteImages}/>);
        wrapper.find("button").simulate("click")
        expect(deleteImages).toHaveBeenCalled();
    });
});

describe("Image list", () => {
    it("should render ImageList", () => {
        const wrapper = shallow(<ImageList images={[]}/>);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Image list container render", () => {
    it("should render ImageListContainer", () => {
            const addImagesToStore = jest.fn();
            const wrapper = shallow(<ImageListContainer images={[]} addImagesToStore={addImagesToStore}/>);
            expect(wrapper).toMatchSnapshot();
        }
    )
    ;
});

describe("Image list container handle scroll", () => {
    it("should handle scroll to bottom in ImageListContainer", () => {
            const addImagesToStore = jest.fn();
            const wrapper = shallow(<ImageListContainer images={[]} addImagesToStore={addImagesToStore}/>);
            const event = {
                target: {
                    scrollingElement: {
                        scrollHeight: 0,
                        clientHeight: 0,
                        scrollTop: 0,
                    }
                }
            }
            wrapper.simulate('scroll', event)
            expect(addImagesToStore).toHaveBeenCalled();
        }
    )
    ;
});

describe("Input Form render", () => {
    it("should render InputForm", () => {
        const wrapper = shallow(<InputForm countToDelete={1} updateCountToDelete={jest.fn}/>);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Input Form onChange", () => {
    it("should handle onChange in InputForm", () => {
        const updateCountToDelete = jest.fn()
        const wrapper = shallow(<InputForm class="a" countToDelete={1} updateCountToDelete={updateCountToDelete}/>);
        const event = {target: {value: 2}}
        wrapper.find("input").simulate('change', event)
        expect(updateCountToDelete).toHaveBeenCalled()

    });
});

describe('Test thunk action when is not fetching', () => {
    it('addCat should start fetching when isFetching=false', () => {
        global.fetch = jest.fn().mockResolvedValueOnce([])
        const func = addCat();
        const dispatch = jest.fn();
        const getState = jest.fn(() => ({isFetching: false}));
        return func(dispatch, getState).then(() => {
            expect(dispatch).toHaveBeenCalledWith({type: START});
            expect(fetch).toHaveBeenCalledWith("https://api.thecatapi.com/v1/images/search?limit=10");
            expect(dispatch).toHaveBeenCalledWith({type: ADD})
            expect(dispatch).toHaveBeenCalledWith({type: END});
        });

    })
})

describe('Test thunk action when is Fetching', () => {
    it('addCat should not start fetching when isFetching=true', () => {
        global.fetch = jest.fn().mockResolvedValueOnce([])
        const func = addCat();
        const dispatch = jest.fn();
        const getState = jest.fn(() => ({isFetching: true}));
        return func(dispatch, getState).then(() => {
            expect(dispatch).not.toHaveBeenCalled();
            expect(fetch).not.toHaveBeenCalled();
        });

    })
})