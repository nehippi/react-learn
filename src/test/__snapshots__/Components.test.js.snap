// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render counter 1`] = `
<p>
  Count of cats: 
  1
</p>
`;

exports[`Delete button render should render Delete button 1`] = `
<button
  onClick={[Function]}
>
  delete 
  1
   cats
</button>
`;

exports[`Image list container render should render ImageListContainer 1`] = `
<ImageList
  data={
    Array [
      Object {
        "id": 1,
        "url": "src/logo.svg",
      },
    ]
  }
/>
`;

exports[`Image list should render ImageList 1`] = `<div />`;

exports[`Input Form render should render InputForm 1`] = `
<fieldset>
  <label>
    Enter count
  </label>
  <input
    min={0}
    onChange={[Function]}
    type="number"
    value={1}
  />
</fieldset>
`;
